[PP] zero-bubble autograd func improvements

H-Huang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134014
* #134052

Remove updates on the autograd graph during zero-bubble input/weight grad updates. Removes the need for weak refs which was causing a memory leak, also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc @XilunWu @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

H-Huang committed:
[pp] zero-bubble autograd func improvements

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


WIP changes to remove updates on the autograd graph during zero-bubble input/weight grad updates. Also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

wconstab commented on test/distributed/pipelining/test_backward.py:
```diff
@@ -117,22 +135,33 @@ def test_stage_backward_weight(self):
+            stage_backward_weight(mod.parameters(), param_groups, retain_graph=True)
+            gc.collect()
+            after_gc_count = len(gc.get_objects())
+            self.assertLessEqual(after_gc_count, before_gc_count)
```
are you monitoring python memory? We probably want to monitor gpu memory and have some tighter bound about the expected size that gets freed

H-Huang committed:
Update on "[PP] zero-bubble autograd func improvements"


Remove updates on the autograd graph during zero-bubble input/weight grad updates. Removes the need for weak refs which was causing a memory leak, also add tests for memory leaks.

TODO:
Need to re-add assert for checking that the param groups and equivalent to equal sizes (https://github.com/pytorch/pytorch/pull/132691#discussion_r1719281924)

cc XilunWu awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]
